pkg/preflight/checkers/checks.go

```diff
@@ -0,0 +1,141 @@
+package checkers
+
+import (
+	"bytes"
+	"fmt"
+	"net"
+	"os"
+	"os/exec"
+	"strings"
+
+	"k8s.io/apimachinery/pkg/util/sets"
+
+	api "github.com/weaveworks/ignite/pkg/apis/ignite"
+	"github.com/weaveworks/ignite/pkg/constants"
+	"github.com/weaveworks/ignite/pkg/providers"
+)
+
+const (
+	oldPathString  = "/"
+	newPathString  = "-"
+	noReplaceLimit = -1
+)
+
+type Checker interface {
+	Check() error
+	Name() string
+	Type() string
+}
+
+type PortOpenChecker struct {
+	port uint64
+}
+
+func (poc PortOpenChecker) Check() error {
+	listener, err := net.Listen("tcp", fmt.Sprintf(":%d", poc.port))
+	if err != nil {
+		return fmt.Errorf("Port %d is in use", poc.port)
+	}
+	if err := listener.Close(); err != nil {
+		return fmt.Errorf("Port %d is in used, failed to close it", poc.port)
+	}
+	return nil
+}
+
+func (poc PortOpenChecker) Name() string {
+	return fmt.Sprintf("Port-%d", poc.port)
+}
+
+func (poc PortOpenChecker) Type() string {
+	return "Port"
+}
+
+type ExistingFileChecker struct {
+	filePath string
+}
+
+func NewExistingFileChecker(filePath string) ExistingFileChecker {
+	return ExistingFileChecker{
+		filePath: filePath,
+	}
+}
+
+func (efc ExistingFileChecker) Check() error {
+	if _, err := os.Stat(efc.filePath); os.IsNotExist(err) {
+		return fmt.Errorf("File %s, does not exist", efc.filePath)
+	}
+	return nil
+}
+
+func (efc ExistingFileChecker) Name() string {
+	return fmt.Sprintf("ExistingFile-%s", strings.Replace(efc.filePath, oldPathString, newPathString, noReplaceLimit))
+}
+
+func (efc ExistingFileChecker) Type() string {
+	return "ExistingFile"
+}
+
+type BinInPathChecker struct {
+	bin string
+}
+
+func (bipc BinInPathChecker) Check() error {
+	_, err := exec.LookPath(bipc.bin)
+	if err != nil {
+		return fmt.Errorf("Bin %s is not in your PATH", bipc.bin)
+	}
+	return nil
+}
+
+func (bipc BinInPathChecker) Name() string {
+	return ""
+}
+
+func (bipc BinInPathChecker) Type() string {
+	return ""
+}
+
+type AvailablePathChecker struct {
+	path string
+}
+
+func StartCmdChecks(vm *api.VM, ignoredPreflightErrors sets.String) error {
+	checks := []Checker{}
+	checks = append(checks, ExistingFileChecker{filePath: "/dev/mapper/control"})
+	checks = append(checks, ExistingFileChecker{filePath: "/dev/net/tun"})
+	checks = append(checks, ExistingFileChecker{filePath: "/dev/kvm"})
+	checks = append(checks, providers.Runtime.PreflightChecker())
```

Comment:
This could be an inline constructor instead of 4 allocations

Reply:
@najeal could you please address this per @twelho comment.